### ISSUE TITLE
[restrict-template-expressions] Errors if type is any

### ISSUE THREAD

**Issue by felixfbecker** [typescript-eslint/typescript-eslint#1660]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/<rule>": ["<setting>"]
  }
}
```

```TS
// your repro code case
try {
} catch (error) {
  throw new Error(`Could not foobar: ${error.message}`)
}
```

**Expected Result**
Should not error, because `any` is assignable to `string` the same way that a string literal type is. (This would be a case though for `no-implicit-any` #791 , but that's not the goal of this rule)

**Actual Result**
"Invalid type of template literal expression."

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `X.Y.Z` |
| `node`                             | `X.Y.Z` |
| `npm`                              | `X.Y.Z` |

**Comment by bradzacher** [comment]
The rule doesn't work off of assignability, it is intended to make sure you're not doing something that might dump garbage into your template strings.

Eg by default the rule errors on `number`, to help prevent overly verbose decimal numbers from leaking into your strings, but there is an option to allow `number` (and `boolean`).

`any` was mentioned on the PR, but never implemented.
happy to accept a new option `allowAny`.

**Comment by felixfbecker** [comment]
I think assignability is what users would expect though. For example, I'd expect a union type `'a' | 'b'` to be allowed, because it is assignable to `string`. Or a type alias `type FilePath = string`. Now whether the implementation actually checks assignability or just does simpler checks that result in the same behavior I think doesn't matter, it's just about the observable behavior.
